pkg/models/job.go

```diff
@@ -114,6 +114,23 @@ type Job struct {
 
 	CreateTime int64 `json:"CreateTime"`
 	ModifyTime int64 `json:"ModifyTime"`
+
+	ReschedulingPolicy ReschedulingPolicy `json:"ReschedulingPolicy,omitempty"`
+}
+
+type ReschedulingPolicy struct {
+	// SchedulingTimeout is the time, in seconds, the job is allowed to wait
+	// to be scheduled before we give up trying.
+	SchedulingTimeout int64 `json:"SchedulingTimeout"`
+
+	// How many seconds to initially wait between retries. This increases each retry, up to MaximumRetryDelay
+	BaseRetryDelay int64 `json:"BaseRetryDelay"`
+
+	// The maximum delay between retries
+	MaximumRetryDelay int64 `json:"MaximumRetryDelay"`
+
+	// The growth factor: how much longer to wait before retrying each time
+	RetryDelayGrowthFactor float64 `json:"RetryDelayGrowthFactor"`
```

Comment:
As I pointed out in our last meeting, how will these timeouts play with the `TimeoutConfig` we have under the task https://github.com/bacalhau-project/bacalhau/blob/3f1733a84d3ae48d61b59f75c19d679fba8745c0/pkg/models/task.go#L40?

`TimeoutConfig` currently only have `ExecutionTimeout` value. While the name might indicate it is a timeout for when an execution start and doesn't cover scheduling or when an execution is rescheduled somewhere else, the implementation is actually an end to end timeout from when the job is first created as shown here https://github.com/bacalhau-project/bacalhau/blob/3f1733a84d3ae48d61b59f75c19d679fba8745c0/pkg/requester/housekeeping.go#L72

The reason we have `TimeoutConfig` under `Task` instead of `Job` is that once we support multiple executions per job, we should allow different timeout values for each task. Also the plan is to add a timeout for pulling the inputs, a timeout for publishing the results, and make execution timeout only for actual execution.

Also, the compute nodes do timeout the execution on their own if it takes longer than `ExecutionTimeout` and report that to the requester. The reason we have a Housekeeper in the Orchestrator side is to make sure lost jobs are eventually marked as failed, such as when the compute node is no longer responsive and didn't report that its execution was timedout to the orchestrator. 

Since we are adding a new `SchedulingTimeout` here, my recommendations are:
- Change Housekeeping to treat execution timeout from when the execution started instead of when the job was created. My guess it that the housekeeper should fail the execution instead of the job and enqueue an evaluation. Then it should be the responsibility of the scheduler to make a decision on whether to retry on another node or not. 
- Clarify what `SchedulingTimeout` is. Is it end to end from the the job was created? Or is it how long the job stays idle in the queue waiting for allocation? I think it makes sense to have both types of timeouts in our config.

It would be nice to do some of these changes on separate PRs as this one keeps getting bigger. Example introduce and merge the housekeeper changes first

Reply:
Yeah, I didn't put this in with the TimeoutConfig because (a) it's about the whole job not just one task and (b) even thought one of the four tunables here *is* a timeout, it's specifically the timeout for retrying scheduling so it seemed nice to put it together with the other things that tune scheduling retries. I think you're right about changing the meaning of the ExecutionTimeout - a user who expects their job to take ten minutes of CPU should be able to set an execution timeout of ten minutes plus a bit of safety margin, and not need to effectively add scheduling time into the execution timeout as well; and a job that happens to get scheduled quickly doesn't deserve to have extra execution time! And that should indeed be a separate change.